### PR TITLE
perf(css/parser): improve lexer perf

### PR DESCRIPTION
**Pull request by alexander-akait**
**Description:**

Here
- benches for lexer/parser
- Implement `Iter` for `TokensInput` and for `Lexer` (now we can remove `next` method in favor iter, I will do it in next PR)
- improve perf 40-50% (based on criterion)

**BREAKING CHANGE:**

No

**Related issue (if exists):**

No

### CODE REVIEW

File: crates/swc_css_parser/src/parser/input.rs [hunk 1]
```diff
@@ -6,7 +6,7 @@ use swc_css_ast::{Token, TokenAndSpan, Tokens};
 use super::PResult;
 use crate::error::{Error, ErrorKind};
 
-pub trait ParserInput {
+pub trait ParserInput: Clone + Iterator<Item = TokenAndSpan> {
     type State: Debug;
 
     fn next(&mut self) -> PResult<TokenAndSpan>;
```
Comment (kdy1) [inline crates/swc_css_parser/src/parser/input.rs:12]: I think we can safely remove this method

File: crates/swc_css_parser/src/parser/input.rs [hunk 2]
```diff
@@ -109,7 +109,7 @@ where
         }
 
         if self.cur.is_none() {
-            let res = self.input.next();
+            let res = ParserInput::next(&mut self.input);
```
Comment (kdy1) [inline crates/swc_css_parser/src/parser/input.rs:112]: So we don't need fully-qualified syntax.

File: crates/swc_css_parser/src/parser/input.rs [hunk 3]
```diff
@@ -6,7 +6,7 @@ use swc_css_ast::{Token, TokenAndSpan, Tokens};
 use super::PResult;
 use crate::error::{Error, ErrorKind};
 
-pub trait ParserInput {
+pub trait ParserInput: Clone + Iterator<Item = TokenAndSpan> {
```
Comment (kdy1) [inline crates/swc_css_parser/src/parser/input.rs:9]: Extending `Iterator` is enough

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc_css_parser/src/parser/input.rs (hunk 1 above)
Comment (alexander-akait) [reply crates/swc_css_parser/src/parser/input.rs:12]: @kdy yes, I write it above, I just want to remove it in other PR + do buffer refactor for perf reasons, so I postpone it

File: crates/swc_css_parser/src/parser/input.rs (hunk 3 above)
Comment (alexander-akait) [reply crates/swc_css_parser/src/parser/input.rs]: Done

File: crates/swc_css_parser/src/lexer/mod.rs [hunk 4]
```diff
@@ -31,13 +35,94 @@ where
...
         let start_pos = input.last_pos();
+
         Lexer {
             input,
             cur: None,
             cur_pos: start_pos,
             start_pos,
             last_pos: None,
             config,
+            buf: Rc::new(RefCell::new(String::with_capacity(256))),
```
Comment (kdy1) [inline crates/swc_css_parser/src/lexer/mod.rs:46]: Do we need `Rc` here?

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - ''

**Pull request merged**